x-pack/plugins/lens/public/shared_components/coloring/color_stops.tsx

```diff
@@ -0,0 +1,285 @@
+        const numberStop = Number(stop);
+        const prevNumberStop = Number(newColorStops[i - 1]?.stop ?? -Infinity);
+        return isValidColor(color) && !Number.isNaN(numberStop) && numberStop >= prevNumberStop;
+      });
+      if (areStopsValid) {
+        onChange(newColorStops.map(({ color, stop }) => ({ color, stop: Number(stop) })));
+      }
+    },
+    [onChange]
+  );
+  const memoizedValues = useMemo(
+    () => colorStops.map(({ color, stop }) => ({ color, stop: String(stop) })),
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [colorStops, reverse]
+  );
+  const { inputValue: localColorStops, handleInputChange: setLocalColorStops } = useDebouncedValue({
+    onChange: onChangeWithValidation,
+    value: memoizedValues,
+  });
+  const [sortedReason, setSortReason] = useState<string>('');
+  const shouldEnableDelete = localColorStops.length > 2;
+
+  const [popoverInFocus, setPopoverInFocus] = useState<boolean>(false);
+
+  // refresh on unmount:
+  // the onChange logic here is a bit different than the one above as it has to actively sort if required
+  const isMounted = useMountedState();
+  useEffect(() => {
+    return () => {
+      if (!isMounted()) {
+        const areStopsValid = localColorStops.every(({ color, stop }, i) => {
+          const numberStop = Number(stop);
+          return isValidColor(color) && !Number.isNaN(numberStop);
+        });
+        const shouldSort = localColorStops.some(({ stop }, i) => {
+          const numberStop = Number(stop);
+          const prevNumberStop = Number(localColorStops[i - 1]?.stop ?? -Infinity);
+          return numberStop < prevNumberStop;
+        });
+        if (areStopsValid && shouldSort) {
+          onChange(
+            localColorStops
+              .map(({ color, stop }) => ({ color, stop: Number(stop) }))
+              .sort(({ stop: stopA }, { stop: stopB }) => Number(stopA) - Number(stopB))
+          );
+        }
+      }
+    };
+  }, [localColorStops, isMounted, onChange]);
+
+  return (
+    <>
+      {sortedReason ? (
+        <EuiScreenReaderOnly>
+          <p aria-live="assertive">
+            {i18n.translate('xpack.lens.dynamicColoring.customPalette.sortReason', {
+              defaultMessage: 'Color stops have been sorted due to new stop value {value}',
+              values: {
+                value: sortedReason,
+              },
+            })}
+          </p>
+        </EuiScreenReaderOnly>
+      ) : null}
+
+      <EuiFlexGroup
+        data-test-subj={`${dataTestPrefix}_dynamicColoring_custom_stops`}
+        direction="column"
+        gutterSize="s"
+      >
+        {localColorStops.map(({ color, stop }, index) => {
+          const prevStopValue = Number(localColorStops[index - 1]?.stop ?? -Infinity);
+          const nextStopValue = Number(localColorStops[index + 1]?.stop ?? Infinity);
+
+          return (
+            <EuiFlexItem
+              key={index}
```

Comment:
As it's possible to change the order of the elements by passing smaller than previous stop value, we shouldn't use index as key. React will be lost and it will mess up the focus.

Reply:
We have this issue in a few other places as well and we use a local UUID as part of the array items which is stripped before calling upstream `setState`.